src/main/java/org/xbill/DNS/SimpleResolver.java

```diff
@@ -444,4 +426,90 @@ private Message sendAXFR(Message query) throws IOException {
   public String toString() {
     return "SimpleResolver [" + address + "]";
   }
+
+  private class VerifyOneResponse {
+    private Message query;
+    private int qid;
+    private Throwable failureReasonException;
+
+    public VerifyOneResponse(Message query, int qid) {
+      this.query = query;
+      this.qid = qid;
+    }
+
+    /**
+     * Keep the exception that caused this to fail. We don't throw the exception immediately
+     * because, when using mDNS, we might get both failing and successful answers, in which case we
+     * just want to discard the bad answers and keep the good ones; the exception only gets thrown
+     * if only bad answers are received.
+     *
+     * @return the Throwable reporting why this parse failed.
+     */
+    public Throwable getFailureReason() {
+      return failureReasonException;
+    }
+
+    public Message parse(byte[] in) {
+      failureReasonException = null;
+
+      // Check that the response is long enough.
+      if (in.length < Header.LENGTH) {
+        failureReasonException = new WireParseException("invalid DNS header - too short");
+        return null;
+      }
+
+      // Check that the response ID matches the query ID. We want
+      // to check this before actually parsing the message, so that
+      // if there's a malformed response that's not ours, it
+      // doesn't confuse us.
+      int id = ((in[0] & 0xFF) << 8) + (in[1] & 0xFF);
+      if (id != qid) {
+        failureReasonException =
+            new WireParseException("invalid message id: expected " + qid + "; got id " + id);
+        return null;
+      }
+
+      Message r;
+      try {
+        r = parseMessage(in);
+      } catch (WireParseException e) {
+        failureReasonException = e;
+        return null;
+      }
+
+      // validate name, class and type (rfc5452#section-9.1)
+      if (!query.getQuestion().getName().equals(r.getQuestion().getName())) {
+        failureReasonException =
+            new WireParseException(
+                "invalid name in message: expected "
+                    + query.getQuestion().getName()
+                    + "; got "
+                    + r.getQuestion().getName());
+        return null;
+      }
+
+      if ((query.getQuestion().getDClass() & ~DClass.UNICAST_RESPONSE)
```

Comment:
- This masking must only be applied for mDNS, not regular DNS
- In the mDNS case, the mask should be 0x7FFF so that 32768 to 65535 is ignored (see the note at [IANA](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-2) or in the RFC)